### PR TITLE
VACMS-15905: Change search error for long string entries

### PR DESCRIPTION
**Pull request by chriskim2311**
## Summary
We have added a search error text change for query's with a max allowable character count of 255. API for search will not be hit if characters are too long logic has been added to prevent this, to help with error reporting.

Other search iterations like R&S redirect to this search page and then runs the search. We are also catching the search from calling the API if the URL search param is too long.

## Related issue(s)

- _Link to ticket created in va.gov-team repo_
department-of-veterans-affairs/va.gov-cms#15905

## Testing done
Tested search query against multiple string lengths. Correct error message displays once the string length requirement is met.

## Screenshots

Search_Results___Veterans_Affairs_desktop

This is from browser mobile view. Actual search bar is above search button on mobile.
Search_Results___Veterans_Affairs_mobile

## What areas of the site does it impact?
Search page

## Acceptance criteria
- [ ] a11y review

### CODE REVIEW

**Review by randimays — COMMENTED** [review]
On mobile, are you able to add some space between the pink area and the text below? I realize this was an existing issue; we can always cut another ticket if it's not a quick fix.

Screenshot 2023-12-04 at 2 02 36 PM

File: src/applications/search/containers/SearchApp.jsx
Comment (chriskim2311) [reply src/applications/search/containers/SearchApp.jsx]: Also changed to a <p>

**Comment by chriskim2311** [comment]
@laflannery Can you take a look at this PR when you get a chance, PR env is building but tomorrow should be ready for a review thanks!

File: src/applications/search/containers/SearchApp.jsx
Comment (maxx1128) [inline src/applications/search/containers/SearchApp.jsx:60]: **Question:** The description says if the query is longer than 255 characters, it should show the error message. But this logic is if the query is as long as, or longer, than 255 characters. Should it be more like `userInput.length > 255`?

**Review by maxx1128 — COMMENTED** [review]
This looks good, I just have one small question.

File: src/applications/search/containers/SearchApp.jsx
Comment (chriskim2311) [reply src/applications/search/containers/SearchApp.jsx:60]: Thanks for the comment Max! My fault on the description, I will update it. It should say max allowable is 255 characters

**Comment by laflannery** [comment]
@chriskim2311 This hadn't gone through a review during design/planning so there are some issues:
1. Can you update the heading in the alert to be an h2? Current it's an h3 which means the heading level structure skips from H1 directly to H3
image
3. The error message is not programmatically connected to the input field which means if a screen reader user comes back to this search results error page and just jumps right into the input field they have no idea there is an error or what it is.
   1. I found this empty status `<span>` in the form and I'm curious what it's for. Could we use this to dynamically input the same error message that would be available for SR users? 
   2. We could add "An error has occurred because the search is over the character limit. Shorten the search and try again."
   3. If that is possible, you would also need to add the ID `search-results-page-dropdown-a11y-status-message` to the `input` element.
4. The "disabled" button is not a very good pattern for accessibility. This button is still available to be activated (with mouse and keyboard), but when folks try nothing happens. This is especially problematic because of the above issue where the error is not connected.
   1. I would like to follow the [USWDS](https://designsystem.digital.gov/components/button/) `aria-disabled` pattern here. When the error is present and the user cannot submit the search form:
   2. Add `aria-disabled="true"` to the button
   3. Style the button to visual show that it is disabled:
      1. background-color: #6F6E6E; color: #f1f1f1; cursor: not-allowed;(the colors maybe need to go through design but these do past color contrast requirements)
image
   4. I don't think we use disabled buttons anywhere (I have an outstanding question to some folks) because we don't have them in our design system. I'm not sure if this would need to get approval from anyone. Might be a question for Jordan or Amanda?
6. Also, any chance you could remove an additional sr-only label I found? Since there is vey clear text that says "Search" having a sr-only label means that this button announces as "Search Search" which is not necessary. If you could remove the use-sr-only that would be nice
image

**Review by pjhill — DISMISSED** [review]

**Comment by chriskim2311** [comment]
@laflannery Thanks for the comments. I have made some changes but
1. Done.
2. Cannot locate the span your talking about in the form. I am using the va-alert component . I can add a new sr-only span and try to include the id for the input when the search string is over the limit.
3. Done. Thanks I confirmed with Jordan to use the design system colors and functionality for a disabled button. I also haven't seen one anywhere but this implementation makes sense here.
4. Cannot find the sr markup in the component for the search button. Is this the one in the page or the one in the header?

**Comment by laflannery** [comment]
@chriskim2311 Sorry I made those screenshots so small, here's a larger one if it helps to locate what I was talking about any easier - both elements are in the form itself on the results page, not the one in the header:
image

Also, note that we are NOT using the component for the search input here, which is why some of the markup is incorrect - if it's possible to change this to use the [DS component](https://design.va.gov/components/search-input) that might be the better choice?

**Comment by jilladams** [comment]
@laflannery if the ask is to change to the component for the entire search form, that's out of scope for this ticket, due to the lift in testing / regression risk, unless we talk more about it with Dave. Let's discuss in planning tomorrow if we can find a little time. cc @FranECross

**Comment by laflannery** [comment]
Accessibility approved with latest updates, Great work!

**Review by maxx1128 — DISMISSED** [review]

File: src/applications/search/styles.scss
Comment (randimays) [inline src/applications/search/styles.scss:198]: Was this a weird linty thing? 🤔

File: src/applications/search/containers/SearchApp.jsx
Comment (randimays) [reply src/applications/search/containers/SearchApp.jsx:60]: I think we may have some wires crossed here; character comparison should be `> 255`; Search.gov allows 255 characters, but not 256.

File: src/applications/search/containers/SearchApp.jsx
Comment (randimays) [inline src/applications/search/containers/SearchApp.jsx:401]: Nitpick **Consider:** we should leave blank lines before `return`s to increase readability.

File: src/applications/search/containers/SearchApp.jsx
Comment (randimays) [inline src/applications/search/containers/SearchApp.jsx:339]: Was this created to handle something we weren't addressing already, like copying/pasting input?

**Review by randimays — CHANGES_REQUESTED** [review]

File: src/applications/search/styles.scss
Comment (chriskim2311) [reply src/applications/search/styles.scss:198]: Changed it back, not sure what happened here

File: src/applications/search/containers/SearchApp.jsx
Comment (chriskim2311) [reply src/applications/search/containers/SearchApp.jsx:339]: This was created to render the correct error message based on input value length, fetching the value from the child component in SearchDropdownComponent

**Comment by randimays** [comment]
@chriskim2311 I just noticed the `search_typeahead` endpoint is still being called if the search input is too long.

**Review by randimays — APPROVED** [review]
LGTM

**Pull request merged**
